File: LeetCode/cpp/zhaohui/1-TwoSum.cpp
#include <iostream>
#include <vector>
#include<algorithm>
using namespace std;
class DataNode {
	int num;
	int ind;
public:
	DataNode(int number,int index) {
		num = number;
		ind = index;
	}
	int getNum() { return num; }
	int getInd() { return ind; }
	
};
bool comp(DataNode data1, DataNode data2) {
	return data1.getNum() < data2.getNum();
}

class Solution {
public:
	vector<int> twoSum(vector<int>& nums, int target) {
		vector<int> index;
		vector<DataNode> data;
		for (unsigned int i = 0; i < nums.size(); i++) {
			data.push_back(DataNode(nums[i], i));
		}
		sort(data.begin(), data.end(), comp);
		int left = 0, right = nums.size() - 1;
		int tempSum = 0;
		while (left < right) {
			tempSum = data[left].getNum() + data[right].getNum();
			if (tempSum == target) {
				if (data[left].getInd() < data[right].getInd()) {
					index.push_back(data[left].getInd());
					index.push_back(data[right].getInd());
				}
				else {
					index.push_back(data[right].getInd());
					index.push_back(data[left].getInd());
				}
			}
			if (tempSum < target) {
				left++;
			}
			else {
				right--;
			}
		}
		return index;
	}

};


int main() {
	Solution solution;
	//int nums[] = { 1,2,3 };
	vector<int> nums = { 1,2,3 };
	//nums.push_back(1);
	int target = 3;
	vector<int> res = solution.twoSum(nums, target);
	for (unsigned int i = 0; i < res.size(); i++) {
		cout << res[i] << endl;
	}
	return 0;
}

File: LeetCode/python/sitao/1_two_sum.py
class Solution:
	# @param {integer[]} nums
	# @param {integer} target
	# @return {integer[]}
	def twoSum(self, nums, target):
		index = sorted(range(len(nums)), key=lambda k:nums[k])
		nums.sort()
		for i in range(0, len(nums)):
			for j in range(i+1, len(nums)):
				if (nums[i] + nums[j]) == target:
					return sorted([index[i]+1, index[j]+1])
				elif (nums[i] + nums[j]) > target:
					break

# numbers = [2, 7, 11, 15]
# target = 9

# numbers = range(0, 32046, 2)
# numbers = eval(open('data.txt').read())
# target = 16021

# numbers = [3,2,4]
# target = 6

numbers = [0, 4, 3, 0]
target = 0

print Solution().twoSum(numbers, target)
